test/functional/p2p_node_network_limited.py

```diff
@@ -46,6 +49,44 @@ def setup_network(self):
         self.add_nodes(self.num_nodes, self.extra_args)
         self.start_nodes()
 
+    def test_avoid_requesting_old_blocks(self):
+        self.log.info("Test node not requesting blocks beyond limited peer threshold")
+        # The request of blocks further than the NETWORK_NODE_LIMITED threshold causes a direct disconnection.
+        pruned_node = self.nodes[0]
+        miner = self.nodes[1]
+        full_node = self.nodes[2]
+
+        # Connect and generate block to ensure IBD=false
+        self.connect_nodes(1, 2)
+        self.connect_nodes(0, 2)
+        self.generate(miner, 1, sync_fun=self.no_op)
+
+        # Verify all peers are at the same height
+        best_block = miner.getbestblockhash()
+        for node in [full_node, pruned_node]:
+            self.wait_until(lambda: node.getbestblockhash() == best_block, timeout=10)
+
+        # Verify that all peers are out of IBD
+        for node in self.nodes:
+            assert not node.getblockchaininfo()['initialblockdownload']
+
+        # Disconnect full_node from the other two peers (the node will remain out of IBD)
+        self.disconnect_nodes(0, 2)
+        self.disconnect_nodes(1, 2)
+        assert_equal(len(full_node.getpeerinfo()), 0)
+
+        # Mine blocks and sync the pruned node. Surpass the NETWORK_NODE_LIMITED threshold.
+        self.generate(miner, MIN_BLOCKS_TO_KEEP + 12, sync_fun=self.no_op)
+        self.sync_blocks([miner, pruned_node])
+
+        # Connect full_node to prune_node and check peers don't disconnect right away.
+        # (they will disconnect if full_node, which is chain-wise behind, request blocks
+        #  older than MIN_BLOCKS_TO_KEEP)
+        self.connect_nodes(0, 2)
+        assert_equal(len(full_node.getpeerinfo()), 1)
+        time.sleep(3)  # Wait for a few seconds to ensure peers remain connected
+        assert_equal(len(full_node.getpeerinfo()), 1)
```

Comment:
This sleep is non-deterministic. The good thing is that it will not cause sporadic test failures, but still, it could occasionally report success even if the full node requests an old block and gets disconnected (false positive). Also it adds unconditional 3 seconds to the tests execution time.

Is it possible to lookup something in the log or via RPC to cross check that the full node has made a decision not to request blocks from the pruned node?

Reply:
Removed the unconditional wait on the last update :).